app/utility/base_knowledge_svc.py

```diff
@@ -0,0 +1,345 @@
+        ###
+        # return facts that are close to supplied fact.
+        #
+        #
+        # Ex:
+        #    - other facts for an agent with given trait/value
+        #    - other facts for the group/agent
+        #    - other facts with same value
+        ###
+        raise NotImplementedError
+
+    def _fact_value_distribution(self, critera, agent=None, group=None):
+        ###
+        # return the value distribution for the given fact, and further filtered down to agent/group if supplied
+        #
+        #
+        # Ex: fact value distribution for 'host.user.name' on group 'workstations':
+        #     --> [{'admin': .4}, {'system': .4}, {'michael': .1}, {'workstation1': .1}]
+        ###
+        raise NotImplementedError
+
+    def _best_guess(self, criteria, agent=None, group=None):
+        # wrapper around 'fact_value_distribution', just returning highest probable value
+        raise NotImplementedError
+
+    def _best_facts(self, agent=None, group=None, metric='usage_success'):
+        ###
+        # best facts based on requested metric
+        #
+        # Args:
+        #    metric: ['usage_success', 'most_recent_success', ...]
+        ###
+        raise NotImplementedError
+
+    def _locate(self, object_name, query=None):
+        """
+        Locate a matching object in the internal store
+        :param object_name: object type
+        :param query: dictionary of fields to match on
+        :return: list of matching objects
+        """
+        try:
+            return [obj for obj in self.fact_ram[object_name] if obj.match(query)]
+        except Exception as e:
+            self.log.error('[!] LOCATE: %s' % e)
+
+    def _remove(self, object_name, query):
+        """
+        Remove objects from the internal store
+        :param object_name: object type
+        :param query: dictionary of fields to match on
+        """
+        try:
+            sublist = [obj for obj in self.fact_ram[object_name] if obj.match(query)]
+            self._clear_matching_constraints(sublist)
+            self.fact_ram[object_name][:] = [obj for obj in self.fact_ram[object_name] if obj not in sublist]
+        except Exception as e:
+            self.log.error('[!] REMOVE: %s' % e)
+
+    def _clear_matching_constraints(self, objs):
+        """
+        Remove constraints associated with objects as part of deletion
+        :param objs: list of objects being removed
+        """
+        need_to_remove = self._get_matching_constraints(objs)
+        for obj in need_to_remove:
+            del self.fact_ram['constraints'][obj._knowledge_id]
+
+    def _get_matching_constraints(self, objs):
+        """
+        Identify matching constraints associated with objects
+        :param objs: list of objects to get constraints for
+        :return: list of relevant constraints
+        """
+        return [obj for obj in objs if obj._knowledge_id in self.fact_ram['constraints']]
```

Comment:
No, you're spot on. This line is wrong since I got some things confused when trying to re-utilize this functionality. I'll fix it (this is kind of related to the missing test you mentioned later on).